### ISSUE TITLE
Question on the parameters speech_threshold and speech_w_music_threshold

### ISSUE THREAD

**Issue by FelixMertin** [NickWilkinson37/voxseg#2]
Hi,

the repo is very easy to use and works as well nicely for german speech! While playing around with your repo and after skimming through the paper, I was wondering about the functionality of the speech_threshold and speech_w_music_threshold parameter. It is clear to me that if you raise them, the activity detection is less sensitive for plain speech and speech in surroundings with music, respectively. My question now is how is the embedment of the threshold working? I had a look into `run_cnnlstm.py` but I am not able to follow you there. Especially this parts is troublesome for me:
```
    prior = np.array([(1-speech_thresh) * speech_w_music_thresh,
                    speech_thresh * speech_w_music_thresh,
                    (1-speech_thresh) * (1-speech_w_music_thresh),
                    (1-speech_thresh) * speech_w_music_thresh])
    temp = pd.concat([_targets_to_endpoints(medfilt([0 if (j*prior).argmax() == 1 else 1 for j in i], filt), 0.32) \
                     for i in targets['predicted-targets']], ignore_index=True)
```

Thanks again for this very useful repo!

**Comment by NickWilkinson37** [comment]
Hi @FelixMertin 

I'm glad to hear the repo is working well for you, and for German too!

The model here is slightly different to the way the model was trained in the paper, I have made a few improvements. In the paper the output layer has a dimension of 2: an output for speech and non-speech, respectively. Here, I trained the model with an output dimension of 4: one output for each of the labels in the AVA-Speech training data ["CleanSpeech", "NoSpeech", "Speech+Music", "Speech+Noise"]. I did this so that I could exclude speech detected within music, which was necessary for one of my use cases (in the radio broadcast domain). This was not possible with the way the model was originally trained.

So what that piece of code is doing, is adjusting the thresholds for each of the different outputs. It is creating a 'mask' to multiply with the output array from the neural net. For example, if the output array was:
```
>>> [0.1, 0.2, 0.6, 0.1]
```
and both thresholds were left at their default value (0.5), the prior array would be:
```
>>> [0.25, 0.25, 0.25, 0.25]
```
When we multiply these arrays the output would be:
```
>>> [0.025, 0.05, 0.15, 0.025]
```
The argmax of this is index 2, which corresponds to the "Speech+Music" label, which means this output will be labeled as speech.

However, if we were to raise the `speech_w_music_threshold` parameter to 0.9, and leave the `speech_threshold` as 0.5, the same example would work out as follows. The output would still be:
```
>>> [0.1, 0.2, 0.6, 0.1]
```
however, the prior array would be:
```
>>> [0.45, 0.45, 0.05, 0.45]
```
When we multiply these arrays the output would be:
```
>>> [0.045, 0.09, 0.03, 0.045]
```
And so now the argmax is index 1, which corresponds to the "NoSpeech" label, which means this output will be labeled as non-speech.

It would probably be easier to follow if I had normalized the prior 'mask' array, so that the default value is [1, 1, 1, 1], however as we are using argmax, it doesn't make a difference, so normalizing would have been unnecessary processing.

The `speech_threshold` works in the same way, and both thresholds may also be used at once, to increase the sensitivity to speech, but decrease the sensitivity to speech within music, or visa versa.

I hope this helps!

**Comment by FelixMertin** [comment]
Hi @NickWilkinson37,

this helps indeed and I'm as well working on a project with radio broadcast sources. Your explanation clears up all of my questions. Thanks!

**Comment by NickWilkinson37** [comment]
No problem @FelixMertin, good luck with your project!

**Comment by adhvaidh1** [comment]
@FelixMertin @NickWilkinson37  I was wondering how you were able to classify the array with decimal values into the 4 labels?
I worked through the entire pipeline, but after classifying into the array, the only option I see is the decode(), which detects speech vs not speech (NAN) at certain points. I would really appreciate the help. Thank you.
